### PR TITLE
Unset function so spy from PATH can run

### LINKED ISSUE #47: Question about nested functions calls
I was looking how to refactor ol’ scripts without too much rewriting and have functions there tested separately. The easiest way was to source target script (with some logic to prevent it running everything) and then do usual spy/asserts for each function.

However, if functionA calls functionB which calls functionC - then spy for functionB is ignored because I think bash first calls sourced functions and then one in PATH.

Workaround is to remove all nested functions with ‘unset -f functionB’ inside each test.

I wonder if that’s something framework can handle, maybe with a flag?

### PR DESCRIPTION
**Pull request by OperationalFallacy**
Hi,

As you pointed out in #47, locally defined functions take precedence over anything in $PATH. unset seems the most straightforward way to fix it, -f needed otherwise it unhappy when there is no function to unset.

Closes #47

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ services:
   ash:
     image: shpy:local
     depends_on: [base]
-    command: sh ${CMD:-/shpy/test/run_tests}
+    command: ash ${CMD:-/shpy/test/run_tests}
```
Comment (codehearts) [inline docker-compose.yml:16]: Thanks for making this change!

**Comment by OperationalFallacy** [comment]
Thank you!

It would be neat to restore functions, but from my limited knowledge of POSIX standard there is no way to print function source in sh/dash (https://www.austingroupbugs.net/view.php?id=1025 there is even a ticket for this). For the rest of shell the approach varies slightly, with typeset and declare build-ins. I've tried to make it work but have to give up quickly after realizing how deep this rabbit hole goes :)

Users who got this far with testing shell scripts to source functions for tests will probably be comfortable with test framework unsetting functions. Maybe leave PR/issue open in case anybody runs into this and find a better solution?

Cheers!

**Comment by codehearts** [comment]
This might actually be possibly with aliases, the following seems to show a way to overwrite local function definitions in a recoverable manner:

```sh
# setup
orig_fn() { echo "original"; }
spy_fn() { echo "spy"; }

orig_fn # original

# alias the local function to the spy function
alias orig_fn=spy_fn

orig_fn # spy

# restore the original alias
unalias orig_fn

orig_fn # original
```

This even allows overwriting and restoring aliases: 🎉 

```sh
# setup
orig_fn() { echo "original"; }
spy_fn() { echo "spy"; }
alias orig_alias=orig_fn

orig_alias # original

# overwrite the alias with the spy
saved_alias="$(alias orig_alias)"
alias orig_alias=spy_fn

orig_alias # spy

# restore the original alias
$saved_alias

orig_alias # original
```

Do you think there'd be any issues with taking this approach? Everything appears to be within the POSIX spec, but I may have missed something

**Comment by OperationalFallacy** [comment]
Interesting idea, I'll try it! I think at some point I tried to use aliases but I couldn't figure out how they work in non-interactive shells and gave up.

**Comment by OperationalFallacy** [comment]
I'm laughing about this, but expanding aliases doesn't work in bash world _if_ it is called inside a function. There is a detailed explanation in https://superuser.com/questions/708462/alias-scoping-in-bash-functions, basically aliases not expanded when function executed - only when it is "read". We have come full circle :)

**Comment by codehearts** [comment]
Lol, I knew that was too good to be true! In that case I've got one last wild idea, but it only seems to work in ash and bash (and zsh if the non-POSIX `functions` is called instead of `set`), and even then it's based on non-POSIX standard behavior. I say just update the docs/readme to mention this shortcoming, and I might cut a ticket later to see if non-standard recovery methods could be added for convenience

```sh
replace_fn() {
  # `set` without arguments outputs all variable and function definitions
  # The range syntax for `sed` can be used to find the start of the function
  # and the final closing brace, allowing the function body to be captured and restored later
  eval "$1_original=\$(set | sed -n '/^$1[[:space:]]*()[[:space:]]*/,/^}[[:space:]]*$/p')"
  eval "$1 () { echo 'replaced'; }"
  export "$1_original"
}

restore_fn() {
  # The first eval fills out the variable name, the second executes its contents
  eval "eval \"\${$1_original}\""
  unset -v "$1_original"
}

foo () { echo 'executing foo'; }
foobar () { echo 'executing foobar'; }
foo
foobar

replace_fn foo
foo
foobar

restore_fn foo
foo
foobar
```

**Comment by OperationalFallacy** [comment]
I see, yeah bash is winning this one :)

The use case for this probably not worth added complexity.

Maybe consider a flag to unset (without restore) in createSpy function? Which is not optimal because tested functions will have to sourced in every test, which could be annoying. 
Test run can also print in stderr a warning: “spy created for xxx won’t run because a function exists. Add -u to the spy definition to unset (remove) function. Original function won’t be restored”

**Comment by codehearts** [comment]
Hmm, I like having a message displayed because more people will see that than the docs. Shpy originally use to define a function for each spy and then unset it when cleaned, so this actually isn't breaking behavior. Just a warning that the original function won't be restored should be good, and an extra flag to unset it sounds like a choice idea!

**Comment by OperationalFallacy** [comment]
I had to set `env bash` in shpy otherwise checkbashisms wasn't happy 

```
checkbashisms_1  | possible bashism in shpy line 292 (type):
checkbashisms_1  |     if type "$1" | grep -q "is a shell function\|^$1 is a function"; then
```

File: README.md [hunk 2]
```diff
@@ -102,6 +102,15 @@ touch my-new-file # outputs "call me once, shame on you" to stdout, returns true
 touch my-new-file # outputs "call me twice, shame on me" to stderr, returns false
 ```
 
+When developing tests for complex functions with long chained calls, source all of them and use spy with -u flag. The flag will unset declared function, so complex functions can be tested with a mix of spies and original functions. 
+
+```sh
+source my_script.sh # contains complex_function that calls file_check and directory_check
+
+createSpy -o 'spy test' -u file_check
+complex_function # complex_function calls file_check, so it will output "spy test". However, without -u createSpy gives a warning that file_check is already declared and complex_function executes declared one instead of spy
```
Comment (codehearts) [inline README.md:111]: Would this be easier to read on multiple lines?

```suggestion
complex_function # complex_function calls file_check, which is mocked to output "spy test"
                 # Without -u, createSpy would warn that file_check is already declared and wouldn't be mocked
```

File: shpy [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env sh
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (OperationalFallacy) [reply shpy]: Nice, with the command -v its a home run! Thanks for all the suggestions, it was fun to find a way of keeping bash in check.

**Pull request merged**